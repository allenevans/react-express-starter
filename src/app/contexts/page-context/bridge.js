/* eslint-disable no-plusplus */
let uniqueId = 0;

const childrenWithKey = children => children.map(child => ({
  ...child,
  key: child.key || `head-${++uniqueId}`,
}));

export class Bridge {
  constructor(context = null) {
    this.context = context;
    this.bodyNodes = [];
    this.headNodes = [];
  }

  /**
   * Add elements to be rendered into the tail of the document body.
   * @param children
   */
  body = (...children) => {
    this.bodyNodes.push(...childrenWithKey(children));
  }

  /**
   * Add elements to be rendered into the head of the html document
   * @param children
   */
  head = (...children) => {
    this.headNodes.push(...childrenWithKey(children));
  }

  get nodes() {
    return {
      body: this.bodyNodes,
      head: this.headNodes,
    };
  }

  statusCode = (code) => {
    this.context.response.status(code);
  }
}
