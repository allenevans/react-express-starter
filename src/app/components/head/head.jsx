import PropTypes from 'prop-types';
import React from 'react';
import { PageContext } from '../../contexts/page-context/page-context';

export const Head = () =>
  <head>
    <meta httpEquiv="X-UA-Compatible" content="IE=edge"/>
    <meta charSet="utf-8"/>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

    <PageContext.Consumer>
      {context => context.nodes.head}
    </PageContext.Consumer>
  </head>;

Head.propTypes = {
  fragment: PropTypes.bool,
};

export default Head;
