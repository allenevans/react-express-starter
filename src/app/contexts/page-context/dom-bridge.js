import { Bridge } from './bridge';

export class DOMBridge extends Bridge {
  body = () => { }

  head = (...children) => {
    const { document } = this.context;

    children.forEach((node) => {
      if (node.type === 'title' && typeof node.props.children === 'string') {
        document.title = node.props.children;
      }
    });
  }

  statusCode = () => {}
}
