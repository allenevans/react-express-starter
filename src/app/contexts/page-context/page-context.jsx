import React from 'react';
import { Bridge } from './bridge';
import { DOMBridge } from './dom-bridge';
import { isNode } from '../../../utils/environment-utils';

const createBridge = context => (isNode() ? new Bridge(context) : new DOMBridge(context || window));

export const PageContext = React.createContext(createBridge());
PageContext.bridge = createBridge;
