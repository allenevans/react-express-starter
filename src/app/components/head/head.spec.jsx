import React from 'react';
import ReactRenderer from 'react-test-renderer';
import { Head } from './head';
import { PageContext } from '../../contexts/page-context/page-context';

describe('head component', () => {
  it('should render the component without the page context', () => {
    const wrapper = ReactRenderer.create(<Head/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the nodes defined in the page context', () => {
    const data = PageContext.bridge();

    data.head(<meta name="meta-test-node" content="META NODE" />);

    const wrapper = ReactRenderer.create(<div>
      <PageContext.Provider value={data}>
        <Head/>
      </PageContext.Provider>
    </div>).root;

    expect(wrapper.findByProps({ name: 'meta-test-node' }).props.content).toEqual('META NODE');
  });
});
