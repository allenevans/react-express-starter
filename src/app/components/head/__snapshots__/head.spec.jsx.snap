// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`head component should render the component without the page context 1`] = `
<head>
  <meta
    content="IE=edge"
    httpEquiv="X-UA-Compatible"
  />
  <meta
    charSet="utf-8"
  />
  <meta
    content="width=device-width, initial-scale=1.0"
    name="viewport"
  />
</head>
`;
